File: dynamic_programming/0309_best_time_to_buy_and_sell_stock_with_colldown/0309_best_time_to_buy_and_sell_stock_with_colldown.go
package main

import (
	"fmt"
)

func maxProfit(prices []int) int {
	n := len(prices)
	if n == 0 {
		return 0
	}

	buy := make([]int, n+1)
	buy[1] = 0 - prices[0]
	sel := make([]int, n+1)

	for i := 2; i <= n; i++ {
		// buy[i] 最佳收益 = max(昨天买入的最佳收益，前天卖出的最佳收益 - 第 i 天价格)
		buy[i] = max(buy[i-1], sel[i-2]-prices[i-1])
		// sel[i] 最佳收益 = max(昨天卖出的最佳收益，昨天买入的最佳收益 + 第 i 天价格)
		sel[i] = max(sel[i-1], buy[i-1]+prices[i-1])
	}
	return sel[n]
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	prices := []int{1, 2, 3, 0, 2}
	res := maxProfit(prices)
	fmt.Println(res)
}

File: dynamic_programming/0309_best_time_to_buy_and_sell_stock_with_colldown/0309_best_time_to_buy_and_sell_stock_with_colldown.py

class Solution(object):
    def maxProfit(self, prices):
        n = len(prices)
        if n < 2:
            return 0
        if n == 2:
            return max(0, prices[n - 1] - prices[n - 2])
        dp_prev_prev = 0
        dp_prev = max(0, prices[n - 1] - prices[n - 2])
        dp_curr = max(dp_prev, prices[n - 2] - prices[n - 3], prices[n - 1] - prices[n - 3])
        max_so_far = max(prices[n - 1], prices[n - 2])
        for i in xrange(n - 4, -1, -1):
            max_so_far = max(max_so_far, prices[i + 1] + dp_prev_prev)
            dp_prev_prev = dp_prev
            dp_prev = dp_curr
            dp_curr = max(dp_curr, max_so_far - prices[i])
        return dp_curr


prices = [1, 2, 3, 0, 2]
res = Solution().maxProfit(prices)
print(res)